x8.md

```diff
@@ -46,7 +47,32 @@ As you can see, the message is an array and each index value has its meaning:
 
 For more details regarding serialization and deserialization please consult [transit format](https://github.com/cognitect/transit-format) specification.
 
-# Content types
+# Message types
+
+- [Message](##Message)
+
+## Message
+
+The type `Message` represents a text message exchanged between clients.
+
+- [Payload](###Payload)
+- [Content types](###Content-types)
+- [Message types](###Message-types)
+- [Clock vs Timestamp and message ordering](###Clock-vs-Timestamp-and-message-ordering)
+- [Replies](###Replies)
+
+### Payload
+
+Payload is a struct (a compound data type) with the following fields (order is importent):
```

Comment:
typo: important

To clarify, this order ensures the hash is deterministic per client implementation above?

Reply:
no,  the hash is deterministic based of sha3 256, 
but the encoding is not.

This means that two different clients, encoding exactly the same message might result in different ids. That's because of protobuf non deterministic encoding (most of the time it is though , there s more on github discussion for protobuf).

In our case this is not a problem, as the hash is always calculated over the encoded version, and no client should decode/re-encode if it needs to propagate/resend, but should always use the unmodified received payload

Reply:
@cammellos, that seems like an accident waiting to happen. I’d rather we pack fields separately so we always have the same hash and don’t hash the encoded message.

Reply:
what do you mean exactly by "packing fields separately" ?

Reply:
HASH(protofield1, ...); otherwise someone will fuck up. Someone always does.

Reply:
> This means that two different clients, encoding exactly the same message might result in different ids. That's because of protobuf non deterministic encoding (most of the time it is though , there s more on github discussion for protobuf).

I believe `**HASH("MESSAGE_ID", group_id, timestamp, body);**` from https://github.com/status-im/bigbrother-specs/blob/master/data_sync/mvds.md would be an example of that. Even though it is still using protobuf

Reply:
I wonder how this would make sense for signatures over arbitrary data though, do we have any other example protocols to look at here?

Also cc @arnetheduck

Reply:
not sure you guys have got this right

Reply:
> Those unable to decrypt the message data are also unable to access the signature. The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity. The signature is serialised as the concatenation of the R, S and V parameters of the SECP-256k1 ECDSA signature, in that order. R and S are both big-endian encoded, fixed-width 256-bit unsigned. V is an 8-bit big-endian encoded, non-normalised and should be either 27 or 28.

From http://eips.ethereum.org/EIPS/eip-627

Seems to indicate field order of underlying data hash is calculated over is important, whereas above seems to indicate it is indeterministic.

As said above, could very well misunderstand something but it seems worthwhile to clarify at least

Reply:
signature should be over encoded data.

Otherwise it makes the protocol non upgradable as you can't add fields, as other clients will not parse that field, and won't use it in the signature.

that s a disaster that will happen.

oskarth what you are suggesting is no different from what i proposed, only we don't have a timestamp, messageid, groupid.


moreover the same problem that you are concerned about happen with that datasync code.

other protocols also calculate signatures over the encoded data for the same reasons (whisper, briar etc)

Reply:
> Those unable to decrypt the message data are also unable to access the signature. The signature, if provided, is the ECDSA signature of the Keccak-256 hash of the unencrypted data using the secret key of the originator identity. The signature is serialised as the concatenation of the R, S and V parameters of the SECP-256k1 ECDSA signature, in that order. R and S are both big-endian encoded, fixed-width 256-bit unsigned. V is an 8-bit big-endian encoded, non-normalised and should be either 27 or 28.

From http://eips.ethereum.org/EIPS/eip-627, where the data appears to be deterministic of data fields.

As said above, could very well misunderstand something but it seems worthwhile to clarify at least

Reply:
> signature should be over encoded data.
>
> Otherwise it makes the protocol non upgradable as you can't add fields, as other clients will not parse that field, and won't use it in the signature.

That's a fair point. I personally find it difficult to evaluate what the consequences are of not being able to get the same signature for the same data, but it could be that this is not something to worry about.

> moreover the same problem that you are concerned about happen with that datasync code.

Why? in MVDS spec it's over some bytes. If clients mess with order of those that's a different question, no?

Reply:
> no client should decode/re-encode if it needs to propagate/resend, but should always use the unmodified received payload

Isn't this adding assumptions on keeping certain payloads around? It at least seems like a trade-off to me, perhaps it's well established what the better solution is (don't care about order in encoding and sign the blob)

Reply:
Ok,
not sure what you are arguing for:

1) In the proposal Hash(protobuf encoded payload)
2) Suggested by dean Hash(Text, ClockValue)

1) is non deterministic because of protobuf, meaning if you encode the same content twice you may get a different hash. Can you make an example of where this is actually going to be a  problem? Just to understand what is the actual concern. The datasync example you sent is exactly this, but we only have a field here (as a matter of fact we should include the pk of the sender when calculating the messageID but that's another story).

2) Is not even worth discussing, as any extra field will break the scheme, and works well only when you have a very well defined set of fields (whisper for example R,S, V , which won't change), but notice that it won't decode the data

We can argue that we should not be using protobuf because of this, but everyone seemed to be onboard and we all knew it was not deterministic.

Is any other proposal worth considering?

Reply:
> Isn't this adding assumptions on keeping certain payloads around? It at least seems like a trade-off to me, perhaps it's well established what the better solution is 

You can decode/encode as fields that you don't know (say a field from a more recent version), will be silently discarded in most implementation, as can't be read.

That means that decoding is **lossy** so re-encoding is not an option.

Reply:
Not arguing for anything specific right now, just trying to explore the problem/solution space. I don't know. By default I do think your proposal is correct, but it still seems worth understanding and elaborating on the consequences a bit more.

Do we have to rush this change? I'd like to hear Jacek's POV on this.

Reply:
> Why? in MVDS spec it's over some bytes. If clients mess with order of those that's a different question, no?

This is the same, the signature is over some bytes, not a specific format. Is on whatever the application layer passes. We could use transit, protobuf or anything, and this code will not change. At the moment we don't even use protobuf, we actually use transit, so it is deterministic. But it should not be a concern here, as the signature code does not have any understanding of what is the content of the data.

> In MVDS, you can see that the fields being hashed are individual protobuf fields, which means determinism is introduced.

This is exactly the same. It's a wrapper, it's not the **application layer** (otherwise it would not be a wrapper). It does not care what the application layer sends. There's no ordering here as there's only **1** field. Namely the payload. So with that respect is deterministic. 


The consequences I indicated is only if we use protobuf for encoding the messages, which currently we don't, we use transit, which afaik is deterministic.

Reply:
Good points, agree. It's probably due to the proximity of the wrapper and application layer that I personally mixed the two